google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionImpl.java

```diff
@@ -476,6 +482,42 @@ public void setTransactionMode(TransactionMode transactionMode) {
     this.unitOfWorkType = UnitOfWorkType.of(transactionMode);
   }
 
+  @Override
+  public String getTransactionTag() {
+    ConnectionPreconditions.checkState(!isClosed(), CLOSED_ERROR_MSG);
+    ConnectionPreconditions.checkState(!isDdlBatchActive(), "This connection is in a DDL batch");
+    return transactionTag;
+  }
+
+  @Override
+  public void setTransactionTag(String tag) {
+    ConnectionPreconditions.checkState(!isClosed(), CLOSED_ERROR_MSG);
+    ConnectionPreconditions.checkState(
+        !isBatchActive(), "Cannot set transaction tag while in a batch");
+    ConnectionPreconditions.checkState(isInTransaction(), "This connection has no transaction");
+    ConnectionPreconditions.checkState(
+        !isTransactionStarted(),
+        "The transaction tag cannot be set after the transaction has started");
+
+    this.transactionBeginMarked = true;
+    this.transactionTag = tag;
+  }
+
+  @Override
+  public String getStatementTag() {
+    ConnectionPreconditions.checkState(!isClosed(), CLOSED_ERROR_MSG);
+    ConnectionPreconditions.checkState(!isDdlBatchActive(), "This connection is in a DDL batch");
+    return statementTag;
+  }
+
+  @Override
+  public void setStatementTag(String tag) {
```

Comment:
Out of curiosity, could we prevent the calling of this for a `COMMIT`? The user should not be able to do a statement tag in that case.

Reply:
Good question. We can't prevent it before the `COMMIT`, but we can throw an exception if `COMMIT` is called after a statement tag has been set, and I think that makes sense as we are quite strict in checking the order of other statements (e.g. you are only allowed to get a commit timestamp if you actually committed etc.).

Reply:
I've changed the implementation to check for this, and to throw an error if the application tries to set a statement tag for a `COMMIT` or `ROLLBACK` statement. This change also adds a change relating to DML batches: DML batches can only include one statement tag, as the statements are sent as one `ExecuteBatchDmlRequest`, which only allows one statement tag. This statement tag must be set before calling `START BATCH DML`, e.g.

```sql
SET STATEMENT_TAG = 'tag-1';
START BATCH DML;
INSERT INTO Singers (SingerId, Name) VALUES (1, 'Morrison');
INSERT INTO Singers (SingerId, Name) VALUES (2, 'Pieterson');
RUN BATCH;
```